File: Sean的leetcode/python/Reverse Vowels of a String.py
###
# Two Pointers
# Time Complexity: O(n) 
# Space Complexity: O(1)
###
class Solution(object):
    def reverseVowels(self, s):
        """
        :type s: str
        :rtype: str
        """
        if len(s) <= 1:
            return s
        l, r = 0, len(s) - 1
        vowels = "AEIOUaeiou"
        s = list(s)
        while l < r:
            if s[l] not in vowels:
                l += 1
            elif s[r] not in vowels:
                r -= 1
            else:
                s[l], s[r] = s[r], s[l]
                l += 1
                r -= 1
        return "".join(s)
                                        
                            
                    

File: MyLeetCode/Leetcode/Reverse Vowels of a String.cpp
//Two pointers
//Time: O(n), Space: O(1)
string reverseVowels(string s) {
    int n = s.size();
    if (n == 0) return "";
    unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
    int l = 0;
    int r = n - 1;
    while (l < r) {
        if (!vowels.count(s[l])) {
            l++;
        }
        else if (!vowels.count(s[r])) {
            r--;
        }
        else {
            swap(s[l], s[r]);
            l++;
            r--;
        }
    }
    return s;
}
//Sean's solution
class Solution {
public:
    string reverseVowels(string s) {
        int f = 0;
        int e = s.size()-1;
        string vowels = "aeiouAEIOU";
        while(f < e){
            if(vowels.find(s[f]) == string::npos){
                f++;
            }
            else if(vowels.find(s[e]) == string::npos){
                e--;
            }
            else{
                char tmp = s[f];
                s[f] = s[e];
                s[e] = tmp;
                f++;
                e--;
            }
        }
        return s;
    
    }
};